// jodit/types/plugins/iframe/config.d 2.ts
/*!
 * Jodit Editor (https://xdsoft.net/jodit/)
 * Released under MIT see LICENSE.txt in the project root for license information.
 * Copyright (c) 2013-2022 Valeriy Chupurnov. All rights reserved. https://xdsoft.net
 */
declare module 'jodit/config' {
    interface Config {
        /**
         * When this option is enabled, the editor's content will be placed in an iframe and isolated from the rest of the page.
         *
         * @example
         * ```javascript
         * Jodit.make('#editor', {
         *    iframe: true,
         *    iframeStyle: 'html{margin: 0px;}body{padding:10px;background:transparent;color:#000;position:relative;z-index:2;\
         *    user-select:auto;margin:0px;overflow:hidden;}body:after{content:"";clear:both;display:block}';
         * });
         * ```
         */
        iframe: boolean;
        /**
         * Allow editing the entire HTML document(html, head)
         */
        editHTMLDocumentMode: boolean;
        /**
         * You can redefine default page
         *
         * @example
         * ```javascript
         * new Jodit('#editor', {
         *    iframe: true,
         *    iframeDefaultSrc: 'http://xdsoft.net/jodit/docs/',
         * });
         * ```
         */
        iframeDefaultSrc: string;
        /**
         * Base URL where the root directory for {@link Jodit.defaultOptions.iframe|iframe} mode
         *
         * @example
         * ```javascript
         * new Jodit('#editor', {
         *    iframe: true,
         *    iframeBaseUrl: 'http://xdsoft.net/jodit/docs/',
         * });
         * ```
         */
        iframeBaseUrl: string;
        /**
         * Iframe title's content
         */
        iframeTitle: string;
        /**
         * Iframe's DOCTYPE
         */
        iframeDoctype: string;
        /**
         * Custom style to be used inside the iframe to display content.
         * @example
         * ```javascript
         * new Jodit('#editor', {
         *    iframe: true,
         *    iframeStyle: 'html{margin: 0px;}',
         * })
         * ```
         */
        iframeStyle: string;
        /**
         * Custom stylesheet files to be used inside the iframe to display content.
         *
         * @example
         * ```javascript
         * new Jodit('#editor', {
         *    iframe: true,
         *    iframeCSSLinks: ['styles/default.css'],
         * })
         * ```
         */
        iframeCSSLinks: string[];
    }
}
export {};
